proto/substrait/algebra.proto

```diff
@@ -228,6 +228,71 @@ message AggregateRel {
   }
 }
 
+// The Window relational operate on a group of rows, referred to as a window, and calculate a return value for each row based on the group of rows.
+message WindowRel {
+  RelCommon common = 1;
+  Rel input = 2;
+  repeated WindowRelFunction window_functions = 3;
+  repeated Expression partition_expressions = 4;
+  repeated SortField sorts = 5;
+
+  substrait.extensions.AdvancedExtension advanced_extension = 10;
+
+  // A window rel function call.
+  message WindowRelFunction {
+    // Points to a function_anchor defined in this plan, which must refer
+    // to a window function in the associated YAML file. Required; 0 is
+    // considered to be a valid anchor/reference.
+    uint32 function_reference = 1;
+
+    // The arguments to be bound to the function. This must have exactly the
+    // number of arguments specified in the function definition, and the
+    // argument types must also match exactly:
+    //
+    //  - Value arguments must be bound using FunctionArgument.value, and
+    //    the expression in that must yield a value of a type that a function
+    //    overload is defined for.
+    //  - Type arguments must be bound using FunctionArgument.type, and a
+    //    function overload must be defined for that type.
+    //  - Enum arguments must be bound using FunctionArgument.enum
+    //    followed by Enum.specified, with a string that case-insensitively
+    //    matches one of the allowed options.
+    repeated FunctionArgument arguments = 9;
+
+    // Options to specify behavior for corner cases, or leave behavior
+    // unspecified if the consumer does not need specific behavior in these
+    // cases.
+    repeated FunctionOption options = 11;
+
+    // Must be set to the return type of the function, exactly as derived
+    // using the declaration in the extension.
+    Type output_type = 7;
+
+    // Defines the record relative to the current record from which the window
+    // extends. The bound is inclusive. If the lower bound indexes a record
+    // greater than the upper bound, TODO (null range/no records passed?
+    // wrapping around as if lower/upper were swapped? error? null?).
+    // Optional; defaults to the start of the partition.
+    Expression.WindowFunction.Bound lower_bound = 5;
+
+    // Defines the record relative to the current record up to which the window
+    // extends. The bound is inclusive. If the upper bound indexes a record
+    // less than the lower bound, TODO (null range/no records passed?
+    // wrapping around as if lower/upper were swapped? error? null?).
+    // Optional; defaults to the end of the partition.
+    Expression.WindowFunction.Bound upper_bound = 4;
+
+    // Defines the window frame: ROWS, RANGE
+    WindowFrame window_frame = 12;
```

Comment:
What is this? Why is this in the function? If it is in a function here, why wouldn't it need to be in the logical function as well? If I can guess at what it is (doing either a sort based op or a hash based op, I would expect it at the operator level since I the purpose of the operator is to do a single thing (single partitioning schema, single sorting scheme).

If I'm an implementer, I have no idea what this means or how to use it.

Reply:
The window_frame is used to define which row to start the window on and where to end it both in [spark ](https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-window.html)and [postgreSQL](https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS).

Reply:
Perhaps call this `BoundsType` or something?  A "window frame" (in Spark) is the entire `{ RANGE | ROWS } { frame_start | BETWEEN frame_start AND frame_end }`.